File: myExpenses/src/withAds/java/org/totschnig/myexpenses/util/ads/PlatformAdHandlerFactory.java
package org.totschnig.myexpenses.util.ads;

import android.content.Context;
import android.view.ViewGroup;

import com.annimon.stream.Stream;

import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.preference.PrefHandler;
import org.totschnig.myexpenses.util.licence.LicenceHandler;

import androidx.annotation.VisibleForTesting;

public class PlatformAdHandlerFactory extends DefaultAdHandlerFactory {

  public PlatformAdHandlerFactory(Context context, PrefHandler prefHandler, String userCountry, LicenceHandler licenceHandler) {
    super(context, prefHandler, userCountry, licenceHandler);
  }

  @Override
  public AdHandler create(ViewGroup adContainer) {
    if (isAdDisabled()) {
      return new NoOpAdHandler(this, adContainer);
    }
    return new AdmobAdHandler(this, adContainer,
        R.string.admob_unitid_mainscreen, R.string.admob_unitid_interstitial);
  }

  @VisibleForTesting
  public AdHandler[] getAdHandlers(ViewGroup adContainer, String adHandler) {
    return Stream.of(adHandler.split(":"))
        .map(handler -> instantiate(handler, adContainer))
        .filter(element -> element != null)
        .toArray(size -> new AdHandler[size]);
  }

  private  AdHandler instantiate(String handler, ViewGroup adContainer) {
    switch (handler) {
      case "Custom": return new AdmobAdHandler(this, adContainer,
          R.string.admob_unitid_custom_banner, R.string.admob_unitid_custom_interstitial);
      case "Ama": return new AmaAdHandler(this, adContainer);
      case "AdMob": return new AdmobAdHandler(this, adContainer,
          R.string.admob_unitid_mainscreen, R.string.admob_unitid_interstitial);
      default: return null;
    }
  }
}

File: myExpenses/src/withAds/java/org/totschnig/myexpenses/util/ads/PlatformAdHandlerFactory.kt
package org.totschnig.myexpenses.util.ads

import android.content.Context
import android.view.ViewGroup
import androidx.annotation.Keep
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.remoteconfig.FirebaseRemoteConfig
import com.google.firebase.remoteconfig.FirebaseRemoteConfigSettings
import org.totschnig.myexpenses.BuildConfig
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.activity.BaseActivity
import org.totschnig.myexpenses.preference.PrefHandler
import org.totschnig.myexpenses.util.crashreporting.CrashHandler
import org.totschnig.myexpenses.util.licence.LicenceHandler
import timber.log.Timber

@Keep
class PlatformAdHandlerFactory(context: Context, prefHandler: PrefHandler, userCountry: String, licenceHandler: LicenceHandler) : DefaultAdHandlerFactory(context, prefHandler, userCountry, licenceHandler) {
    override fun create(adContainer: ViewGroup, baseActivity: BaseActivity): AdHandler {
        val adHandler = if (isAdDisabled) "NoOp" else {
            val remoteConfig = FirebaseRemoteConfig.getInstance()
            if (BuildConfig.DEBUG) {
                val configSettings = FirebaseRemoteConfigSettings.Builder()
                        .setMinimumFetchIntervalInSeconds(0)
                        .build()
                remoteConfig.setConfigSettingsAsync(configSettings)
            }
            remoteConfig.setDefaultsAsync(R.xml.remote_config_defaults)
            remoteConfig.fetchAndActivate().addOnCompleteListener { task ->
                if (!task.isSuccessful) {
                    task.exception?.let {
                        CrashHandler.report(it)
                    } ?: run {
                        Timber.d("Firebase Remote Config Fetch failed")
                    }
                }
            }
            remoteConfig.getString("ad_handling_waterfall").split(":".toRegex()).getOrNull(0) ?: "AdMob"
        }
        FirebaseAnalytics.getInstance(context).setUserProperty("AdHandler", adHandler)
        return instantiate(adHandler, adContainer, baseActivity)
    }

    private fun instantiate(handler: String, adContainer: ViewGroup, baseActivity: BaseActivity): AdHandler {
        return when (handler) {
            "Custom" -> AdmobAdHandler(this, adContainer, baseActivity,
                    R.string.admob_unitid_custom_banner, R.string.admob_unitid_custom_interstitial, 0)
            "AdMob" -> AdmobAdHandler(this, adContainer, baseActivity,
                    R.string.admob_unitid_mainscreen, R.string.admob_unitid_interstitial, R.string.admob_unitid_interstitial_smart_segmentation)
            else -> NoOpAdHandler
        }
    }
}